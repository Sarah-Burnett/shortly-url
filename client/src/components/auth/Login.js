import React, { useState } from 'react';
import { Redirect } from 'react-router-dom';
import jwt_decode from "jwt-decode";
import { setAuthToken }  from '../../utilities/setAuthToken';
import Errors from './AuthErrors';
import axios from 'axios';
import { isLoginValid } from '../../utilities/validateAuth';
import { AuthForm, Label, Input, FormBtn, PrimaryFormBtn } from '../Styles';
import { useUser } from '../../hooks/useUser';
import { useLinks } from '../../hooks/useLinks';

const Login = () => {
    const { setUser } = useUser();
    const { removeBrowserLinks } = useLinks();
    const [ formValues, setFormValues ] = useState({
        email: '',
        password: '',
    });
    const [ errors, setErrors ] = useState([]);
    const { email, password} = formValues;
    const handleChange = (event) => {
        event.persist()
        setFormValues(values => ({...values, [event.target.id]: event.target.value}))   
    };
    const login = (event) => {
        event.preventDefault();
        const errorMsgs = isLoginValid(formValues);
        if (errorMsgs.length !== 0) return setErrors(errorMsgs);
        axios
          .post('/api/users/login', {email, password})
          .then((res) => {
              const { token } = res.data;
              const { firstName, id } = jwt_decode(token);
              setAuthToken(token);
              setUser(() => ({ isAuth: true, firstName, id, token}));
              removeBrowserLinks();
              return <Redirect to="/dashboard"/>
          })
          .catch(err => setErrors([err.response.data.error || "Login failed. Please try again"]));
    };
  
    return (
        <AuthForm onSubmit={login} noValidate>
            <h3>Login</h3>
            { errors.length > 0 && <Errors errors={errors}/> }
            <Label HTMLfor="email">Email <Input type="email" id="email" name="email" onChange={handleChange} value={email} required></Input></Label>
            <Label HTMLfor="password">Password <Input type="password" id="password" name="password" onChange={handleChange} value={password}required></Input></Label>
            <PrimaryFormBtn type="submit">Submit</PrimaryFormBtn>
            <FormBtn type="button">Forgot password</FormBtn>
        </AuthForm>
    )
};

export default Login;