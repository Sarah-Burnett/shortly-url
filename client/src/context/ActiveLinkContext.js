import React, { useState, createContext } from 'react';

export const ActiveLinkContext = createContext();

export const ActiveLinkProvider = (props) => {
    const [ activeLink, setActiveLink] = useState({
        activeLink: [],
        process: false
    });

    return (
        <ActiveLinkContext.Provider value={[activeLink, setActiveLink]}>
            {props.children}
        </ActiveLinkContext.Provider>
    )
};