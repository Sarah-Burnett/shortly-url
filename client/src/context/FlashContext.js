import React, { useState, createContext } from 'react';

export const FlashContext = createContext();

export const FlashProvider = (props) => {
    const [ flash, setFlash ] = useState({
        flashMsg: "",
        isFlashVisible: false,
        isAutoHideFlash: true,
    });
    return (
        <FlashContext.Provider value={[flash, setFlash]}>
            {props.children}
        </FlashContext.Provider>
    )
};