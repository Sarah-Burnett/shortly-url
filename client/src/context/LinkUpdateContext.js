import React, { useState, createContext } from 'react';

export const LinkUpdateContext = createContext();

export const LinkUpdateProvider = (props) => {
    const [ linkUpdate, setLinkUpdate] = useState();
    return (
        <LinkUpdateContext.Provider value={[linkUpdate, setLinkUpdate]}>
            {props.children}
        </LinkUpdateContext.Provider>
    )
};