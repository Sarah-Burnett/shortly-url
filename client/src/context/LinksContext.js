import React, { useState, createContext } from 'react';

export const LinksContext = createContext();

export const LinksProvider = (props) => {
    const [ links, setLinks] = useState([]);

    return (
        <LinksContext.Provider value={[links, setLinks]}>
            {props.children}
        </LinksContext.Provider>
    )
};