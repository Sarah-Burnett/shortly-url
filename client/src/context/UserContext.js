import React, { useState, createContext } from 'react';

export const UserContext = createContext();

export const UserProvider = (props) => {
    const [ user, setUser] = useState({
        isAuth: false,
        firstName: undefined,
        id: undefined,
        token: undefined,
    });
    
    return (
        <UserContext.Provider value={[user, setUser]}>
            {props.children}
        </UserContext.Provider>
    )
};