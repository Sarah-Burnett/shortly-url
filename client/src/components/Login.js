import React, { useState, useContext } from 'react';
import { Redirect } from 'react-router-dom';
import jwt_decode from "jwt-decode";
import { PrimaryFormBtn, FormBtn, Form, Label, Input } from '../Styles';
import { UserContext } from '../context/UserContext';
import { LinksContext } from '../context/LinksContext';
import { setAuthToken }  from '../utilities/setAuthToken';
import { Errors } from './AuthErrors';
import axios from 'axios';

export const Login = () => {
    const [ user, setUser ] = useContext(UserContext);
    const [ links, setLinks ] = useContext(LinksContext);

    const [ formValues, setFormValues ] = useState({
        email: '',
        password: '',
    });

    const handleChange = (event) => {
        event.persist()
        setFormValues(values => ({...values, [event.target.id]: event.target.value}))   
    };

    const login = (event) => {
        event.preventDefault()
        const { email, password } = formValues;
        axios
          .post('/api/users/login', {email, password})
          .then((res) => {
              const { token } = res.data;
              localStorage.setItem("jwtToken", token);
              setAuthToken(token);
              const {firstName, id} = jwt_decode(token);
              setUser(() => ({ isAuth: true, firstName, id, token}));
              setLinks([]);
              return <Redirect to="/dashboard"/>
          })
          .catch(err => setFormValues(values => ({...values, errors: err})))
    };

    const { email, password, errors } = formValues;

    return (
        <Form onSubmit={login}>
            <h3>Login</h3>
            { formValues.errors && <Errors errors={errors}/> }
            <Label HTMLfor="email">Email <Input type="email" id="email" name="email" onChange={handleChange} value={email} required></Input></Label>
            <Label HTMLfor="password">Password <Input type="password" id="password" name="password" onChange={handleChange} value={password}required></Input></Label>
            <PrimaryFormBtn type="submit">Submit</PrimaryFormBtn>
            <FormBtn type="button">Forgot password</FormBtn>
        </Form>
    )
}
