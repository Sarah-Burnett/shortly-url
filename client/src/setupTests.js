import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, screen, waitForElement } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer';
import { UserProvider } from './context/UserContext';
import { LinksProvider } from './context/LinksContext';
import { ActiveLinkProvider } from './context/ActiveLinkContext';
import { ProcessLinkProvider } from './context/ProcessLinkContext';
import { FlashProvider } from './context/FlashContext';
import App from './App';

export const ComponentWrapper = (Component) => {
    return (
        <UserProvider>
            <LinksProvider>
                <ActiveLinkProvider>
                    <ProcessLinkProvider>
                        <FlashProvider>
                            <Component/>
                        </FlashProvider>
                    </ProcessLinkProvider>
                </ActiveLinkProvider>
            </LinksProvider>
        </UserProvider>
    )
}

export const rendersWithoutCrashing = (Component) => {
    const div = document.createElement('div');
    ReactDOM.render(<Component/>, div);
}

export const matchesSnapshot = (Component) => {
    const tree = renderer.create(<Component/>).toJSON();
    expect(tree).toMatchSnapshot();
}

export const sampleLinks = [
    {
        "date":"2020-08-14T17:29:53.041Z",
        "click":[
            {"date":1597426302926},
            {"date":1597426306071},
            {"date":1597426310231},
            {"date":1597426314763},
            {"date":1597428206995},
            {"date":1597428210093},
            {"date":1597428215131},
            {"date":1597428218755}
        ],
        "_id":"K1pjMCK-F",
        "full":"http://reactcommunity.org/react-transition-group/css-transition",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"octopus","__v":8
    },
    {
        "date":"2020-08-14T17:33:55.748Z",
        "click":[],
        "_id":"lsncCe5RV",
        "full":"https://ba-grant.netlify.app/",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"bagrant","__v":0
    }
];
