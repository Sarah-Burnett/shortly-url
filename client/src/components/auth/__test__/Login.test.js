import React from "react";
import { rendersWithoutCrashing, matchesSnapshot } from "../../../setupTests";
import {
	render,
	cleanup,
	fireEvent,
	screen,
	waitFor,
} from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import Login from "../Login";
import { UserProvider } from "../../../context/UserContext";
import { LinksProvider } from "../../../context/LinksContext";
import { FlashProvider } from "../../../context/FlashContext";

const LoginWrapper = (formValues) => {
	return (
		<UserProvider>
			<LinksProvider>
				<FlashProvider>
					<Login
						formValues={formValues}
						setFormValues={() => {}}
						loginUrl={"/login"}
					/>
				</FlashProvider>
			</LinksProvider>
		</UserProvider>
	);
};

const ValidLoginWrapper = () =>
	LoginWrapper({
		email: "sarah@gmail.com",
		password: "123456",
	});

const InvalidLoginWrapper = () =>
	LoginWrapper({
		email: "",
		password: "",
	});

const server = setupServer(
	rest.post("/login", (req, res, ctx) => {
		console.log("msw");
		return res(
			ctx.json({
				id: "sdfaskldfjasdjfa;fj",
				token:
					"Bearer eyJhbGciOadfsdfsd5cCI6IkpXVCJ9.eyJmaXJzdEasfdsafasdfImlkIjoiNWViZTasfdkMjM1ZTRhNTAxZWU4Njk5MDU4IiwiaWF0IjoxNTk3MDc0OTMwLCJleHAiOjE1OTk2NjY5MzB9.AzE-5eGRAnX_WQCz4gOxeZdgqaG9FSt5__mvB-5XevY",
			})
		);
	})
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterEach(cleanup);
afterAll(() => server.close());

//TODO:
// test 404 server error - err.response.data.error undefined

it("renders login without crashing", () =>
	rendersWithoutCrashing(ValidLoginWrapper));

it("login matches snapshot", () => matchesSnapshot(ValidLoginWrapper));

it("valid user is logged in", async () => {
	render(<ValidLoginWrapper />);
	await fireEvent.click(screen.getByText("Submit"));
	expect(screen.queryByTestId("errors")).toBe(null);
});

it("invalid user shows error", async () => {
	render(<InvalidLoginWrapper />);
	fireEvent.click(screen.getByText("Submit"));
	await waitFor(() => screen.getByTestId("errors"));
	expect(screen.getByTestId("errors")).toHaveTextContent(
		"Please enter your emailPlease enter your password"
	);
});

it("handles server error", async () => {
	server.use(
		rest.post("/login", (req, res, ctx) => {
			return res(ctx.status(500));
		})
	);
	render(<ValidLoginWrapper />);
	fireEvent.click(screen.getByText("Submit"));
	await waitFor(() => screen.getByTestId("errors"));
	expect(screen.getByTestId("errors")).toHaveTextContent(
		"Login failed. Please try again"
	);
});
