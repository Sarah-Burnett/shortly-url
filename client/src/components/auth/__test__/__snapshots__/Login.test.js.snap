// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login matches snapshot 1`] = `
<form
  className="sc-AxhUy dLXkxw"
  data-testid="login"
  noValidate={true}
  onSubmit={[Function]}
>
  <h3>
    Login
  </h3>
  <label
    className="sc-Axmtr fYuxyA"
  >
    Email
     
    <input
      className="sc-AxmLO erTjTr"
      id="email"
      name="email"
      onChange={[Function]}
      required={true}
      type="email"
      value="sarah@gmail.com"
    />
  </label>
  <label
    className="sc-Axmtr fYuxyA"
  >
    Password
     
    <input
      className="sc-AxmLO erTjTr"
      id="password"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value="123456"
    />
  </label>
  <button
    className="sc-AxirZ sc-AxheI cOblFm"
    type="submit"
  >
    Submit
  </button>
  <button
    className="sc-AxjAm sc-AxgMl bwgmuq"
    type="button"
  >
    Forgot password
  </button>
</form>
`;
